programs/showhostkey/showhostkey.8
'\" t
.\"     Title: IPSEC_SHOWHOSTKEY
.\"    Author: Paul Wouters
.\" Generator: DocBook XSL Stylesheets v1.77.1 <http://docbook.sf.net/>
.\"      Date: 12/16/2012
.\"    Manual: Executable programs
.\"    Source: libreswan
.\"  Language: English
.\"
.TH "IPSEC_SHOWHOSTKEY" "8" "12/16/2012" "libreswan" "Executable programs"
.\" -----------------------------------------------------------------
.\" * Define some portability stuff
.\" -----------------------------------------------------------------
.\" ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
.\" http://bugs.debian.org/507673
.\" http://lists.gnu.org/archive/html/groff/2009-02/msg00013.html
.\" ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
.ie \n(.g .ds Aq \(aq
.el       .ds Aq '
.\" -----------------------------------------------------------------
.\" * set default formatting
.\" -----------------------------------------------------------------
.\" disable hyphenation
.nh
.\" disable justification (adjust text to left margin only)
.ad l
.\" -----------------------------------------------------------------
.\" * MAIN CONTENT STARTS HERE *
.\" -----------------------------------------------------------------
.SH "NAME"
ipsec_showhostkey \- show host\*(Aqs authentication key
.SH "SYNOPSIS"
.HP \w'\fBipsec\fR\ 'u
\fBipsec\fR \fIshowhostkey\fR [\-\-ipseckey] [\-\-left] [\-\-right] [\-\-dump] [\-\-verbose] [\-\-version] [\-\-list] [\-\-gateway\ \fIgateway\fR] [\-\-precedence\ \fIprecedence\fR] [\-\-dhclient] [\-\-file\ \fIsecretfile\fR] [\-\-keynum\ \fIcount\fR] [\-\-id\ \fIidentity\fR]
.SH "DESCRIPTION"
.PP
\fIShowhostkey\fR
outputs (on standard output) a public key suitable for this host, in the format specified, using the host key information stored in
/etc/ipsec\&.secrets\&. In general only the super\-user can run this command, since only he can read
\fIipsec\&.secrets\fR\&.
.PP
The
\fB\-\-left\fR
and
\fB\-\-right\fR
options cause the output to be in
\fBipsec.conf\fR(5)
format, as a
\fBleftrsasigkey\fR
or
\fBrightrsasigkey\fR
parameter respectively\&. Generation information is included if available\&. For example,
\fB\-\-left\fR
might give (with the key data trimmed down for clarity):
.sp
.if n \{\
.RS 4
.\}
.nf
  # RSA 2048 bits   xy\&.example\&.com   Sat Apr 15 13:53:22 2000
  leftrsasigkey=0sAQOF8tZ2\&.\&.\&.+buFuFn/
.fi
.if n \{\
.RE
.\}
.PP
The
\fB\-\-ipseckey\fR
option causes the output to be in opportunistic\-encryption DNS IPSECKEY record format (RFC 4025)\&. A gateway can be specified with the
\fB\-\-gateway\fR, which currently supports IPv4 and IPv6 addresses\&. The host name is the one included in the key information (or, if that is not available, the output of
\fBhostname\ \&\-\-fqdn\fR), with a
\fB\&.\fR
appended\&. For example,
\fB\-\-ipseckey \-\-gateway 10\&.11\&.12\&.13\fR
might give (with the key data trimmed for clarity):
.sp
.if n \{\
.RS 4
.\}
.nf
      IN    IPSECKEY  10 1 2 10\&.11\&.12\&.13  AQOF8tZ2\&.\&.\&.+buFuFn/"
.fi
.if n \{\
.RE
.\}
.PP
The
\fB\-\-version\fR
option causes the version of the binary to be emitted, and nothing else\&.
.PP
The
\fB\-\-verbose\fR
may be present one or more times\&. Each occurance increases the verbosity level\&.
.PP
The
\fB\-\-dhclient\fR
option cause the output to be suitable for inclusion in
\fBdhclient.conf\fR(5)
as part of configuring WAVEsec\&. See <\m[blue]\fBhttp://www\&.wavesec\&.org\fR\m[]>\&.
.PP
Normally, the default key for this host (the one with no host identities specified for it) is the one extracted\&. The
\fB\-\-id\fR
option overrides this, causing extraction of the key labeled with the specified
\fIidentity\fR, if any\&. The specified
\fIidentity\fR
must
\fIexactly\fR
match the identity in the file; in particular, the comparison is case\-sensitive\&.
.PP
There may also be multiple keys with the same identity\&. All keys are numbered based upon their linear sequence in the file (including all include directives)
.PP
The
\fB\-\-file\fR
option overrides the default for where the key information should be found, and takes it from the specified
\fIsecretfile\fR\&.
.SH "DIAGNOSTICS"
.PP
A complaint about \(lqno pubkey line found\(rq indicates that the host has a key but it was generated with an old version of FreeS/WAN and does not contain the information that
\fIshowhostkey\fR
needs\&.
.SH "FILES"
.PP
/etc/ipsec\&.secrets
.SH "SEE ALSO"
.PP
ipsec\&.\fBsecrets\fR(5), ipsec\&.\fBconf\fR(5),
\fBipsec_rsasigkey\fR(8)
.SH "HISTORY"
.PP
Written for the Linux FreeS/WAN project <\m[blue]\fBhttp://www\&.freeswan\&.org\fR\m[]> by Henry Spencer\&. Updated by Paul Wouters for the IPSECKEY format\&.
.SH "BUGS"
.PP
Arguably, rather than just reporting the no\-IN\-KEY\-line\-found problem,
\fIshowhostkey\fR
should be smart enough to run the existing key through
\fIrsasigkey\fR
with the
\fB\-\-oldkey\fR
option, to generate a suitable output line\&.
.PP
The
\fB\-\-id\fR
option assumes that the
\fIidentity\fR
appears on the same line as the
\fB:\ \&RSA\ \&{\fR
that begins the key proper\&.
.SH "AUTHOR"
.PP
\fBPaul Wouters\fR
.RS 4
placeholder to suppress warning
.RE


CHANGES

See also https://libreswan.org/

v3.0 (unreleased)
* FORK: Rename from Openswan to Libreswan [Team]
        (for older CHANGES see docs/CHANGES.openswan)
* FORK: Changed our VendorID prefix to "OEN" [Team]
* LICENSE: Updated FSF address on the GPLv2 COPYING file [Team]
* TRADEMARK: Give everyone unlimitel eternal royalty-free license to
        use the name "libreswan" to refer to this software and website [Team]
* NSS: is now mandatory - custom crypto code removed [Paul]
* NSS: Support reading NSS password from file [Avesh]
* NSS: Added "ipsec initnss" and "ipsec import" commands [Paul]
* NSS: We need to include nsperror.h for PR_GetError() [Paul]
* NSS: PK11_DigestFinal() passed sizeof pointer instead of passing
        sizeof *pointer [Paul]
* NSS: use pkg-config to find the right cflags and libs [Paul]
* DNS: Removed LWRES code and old static ISC libraries [Paul]
* DNS: Don't attempt to resolve numerical sourceip= values [Paul]
* DNS: starter and pluto now support USE_DNSSEC using libunbound [Paul]
* OE: Removed support for old KEY and TXT DNS records [Paul]
* OE: Add support for IPSECKEY in ipsec showhostkey [Paul]
* pluto: --config <fn> uses libipsecconf to read 'config setup' [Kim B. Heino]
* pluto: left=%defaultroute now obtains src ip from routing table [Kim B. Heino]
* pluto: Removed support for non-strict ike/esp lines [Paul]
* pluto: UDPFROMTO support was not enabled for NETKEY if not also build
         with KLIPS [Paul]
* pluto: Pass traffic selectors to the kernel in Transport Mode [Avesh]
         (rhbz#831669)
* pluto: Fix phase confusion in xauth/modeconfig [Avesh]
* pluto: Added new option plutostderrlogtime= (default=no) [Paul]
* pluto: Additional safety checks to strncat() calls for addrtot(),
         inet_addrtot(), sin_addrtot(), alg_info_snprint_esp(),
         alg_info_snprint_ah(), idtoa() and format_end() [Paul]
* pluto: Removed unused OCSP code [Paul]
* pluto: crlcheckinterval did not interpret plain numbers as seconds [Philippe]
* pluto: Change ft_mbz to ft_zig - Don't error on "must be zero" but instead
         "zero ignore". This works around an Android ICS/racoon bug [Paul]
* pluto: Update known vendorids [Paul]
* pluto: phased out HAVE_THREADS, pluto/pam now thread-safe [Philippe/Paul]
* pluto: Fixed IPSEC_CONFDDIR handling which broke NSS in tests [Paul]
* pluto: obsoleted prepluto= postpluto= plutoopts= config setup options [Paul]
* pluto: obsoleted plutowait= and pluto= config setup option [Paul]
* pluto: obsoleted nocrsend= option removed (use leftsendcert=) [Paul]
* pluto: removed manual keying remnants [Paul]
* pluto: remove protostack=auto and --use-auto, netkey is new default [Paul]
* pluto: Added perpeerlog=yes|no and perpeerlogdir=/var/log [Paul]
* pluto: Added retransmits=yes|no (matches pluto --noretransmits) [Paul]
* pluto: Added plutofork=yes|no to match pluto --nofork [Paul]
* pluto: added ikeport= and nat_ikeport= options, and --natikeport [Paul]
* pluto: support for secretsfile= and ipsecdir= in ipsec.conf [Paul]
* pluto: remove old unsused USE_IPSECPOLICY code [Paul]
* pluto: rhbz#609343: pluto crashes when removing logical interface [Avesh]
* pluto: dont stop processing after --coredir argument [Paul]
* pluto: perform whack --listen and addconn --autoall on startup [Paul]
* pluto: if started with --nofork, don't care about existing pid file [Paul]
* pluto: incorrect free in scan_proc_shunts() [Roel van Meer]
* pluto: eclipsed() was broken since freeswan-2.02 [Philippe]
* _plutoload: obsoleted [Kim/Paul]
* auto: no longer pass defaultroute/defaultrouteaddr to addconn [Paul]
* whack: fix handling --sha2_truncbug and --nm_configured options [Paul]
* whack: don't try to write to closed stdout tty [Philippe]
* DPD: reduce flood of DPD messages with unexpected seqno [Andrey Alexandrenko]
* DPD: We did not send DPD VID in aggressive mode with NAT-T disabled
* DPD: dpdaction=restart can cause full phase1 timeout after DPD 
  (rhbz#848132) [Avesh]
* PAM: updated contrib/pam.d/pluto (rhbz#815127) [Philippe Vouters]
* PAM: move pam out of contrib, and install config when HAVE_XAUTHPAM [Paul]
* IKEv1: In aggresive mode: allow ISAKMP_NEXT_CR ISAKMP_NEXT_CERT as
         payloads [Philippe]
* IKEv1: aggressive mode sometimes picked wrong RSA/PSK conn [Philippe]
* IKEv1: Simplify outgoing NAT-T proposals, fix logging [Paul]
* XAUTH: Support for runtime choice of xauthby=<pam|file> [Philippe]
* XAUTH: Support for Mutual RSA + XAuth (interop with Shrew Soft) [Philippe]
* XAUTH: Fixed updown to remove ModeCfg (cisco) obtained sourceip [Avesh/Tuomo]
* XAUTH: Do not redo xauth/modecfg during rekey to cisco [Avesh]
* XAUTH: Use incoming XAUTH VID when picking best connection [Philippe]
* XAUTH: pam was failing when built with USE_LIBCAP_NG=true [Philippe Vouters]
* XAUTH: Fixup of defines [Paul/Philippe]
* XAUTH: Don't use XAUTH VID to put conn in policy XAUTH [Andrey Alexandrenko]
* XAUTH: Fix XAUTH TYPE handling and logging [Philippe]
* IKEv2: Comply to RFC's for "must be zero" to ignore instead of abort [Paul]
         (rhbz#831669)
* IKEv2: road warrior support [Antony/Paul/Avesh]
* IKEv2: narrowing code extended to cover ports,protocol,subnets [Antony/Paul]
* Only set MODP768_MODULUS with USE_VERYWEAK_DH1 [Paul]
* NETKEY: ignore interfaces= line for NETKEY [Paul]
* NETKEY: Fix for three AES-GCM issues with key lengths 128, 192, 256 bits
          and IV of 8, 12, 16 bytes as per RFC 4106 [Avesh]
* NETKEY: Labeled IPsec updates [Avesh]
* NETKEY: Support for SHA384/SHA512 and integ(ikev2)  in ESP [Avesh]
* NETKEY: In _updown.netkey, insert route on correct interface when nexthop
          is used [Tuomo]
* NETKEY: Revert "Always use XFRM_MSG_UPDPOLICY instead of XFRM_MSG_NEWPOLICY"
          This caused module unload issues and XFRM_MSG_REPLACE errors [Paul]
* KLIPS: Removed support for Linux < 2.4.4 [Paul]
* KLIPS: Changed _startklips to use ip route instead of netstat [Harald]
* KLIPS: misc. fixes, mostly satot() related [David]

programs/showhostkey/showhostkey.c
/*
 * show the host keys in various formats
 * Copyright (C) 2005 Michael Richardson <mcr@xelerance.com>
 * Copyright (C) 1999, 2000, 2001  Henry Spencer.
 * Copyright (C) 2003-2008 Michael C Richardson <mcr@xelerance.com> 
 * Copyright (C) 2003-2010 Paul Wouters <paul@xelerance.com> 
 * Copyright (C) 2009 Avesh Agarwal <avagarwa@redhat.com>
 * Copyright (C) 2009 Stefan Arentz <stefan@arentz.ca>
 * Copyright (C) 2010 Tuomo Soini <tis@foobar.fi>
 * Copyright (C) 2012 Paul Wouters <pwouters@redhat.com> 
 * Copyright (C) 2012 Paul Wouters <paul@libreswan.org>
 *
 * This program is free software; you can redistribute it and/or modify it
 * under the terms of the GNU General Public License as published by the
 * Free Software Foundation; either version 2 of the License, or (at your
 * option) any later version.  See <http://www.fsf.org/copyleft/gpl.txt>.
 * 
 * This program is distributed in the hope that it will be useful, but
 * WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY
 * or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
 * for more details.
 *
 * replaces a shell script.
 *
 */

#include <sys/types.h>
#include <sys/stat.h>
#include <stdio.h>
#include <limits.h>
#include <errno.h>
#include <string.h>
#include <stdlib.h>
#include <unistd.h>
#include <getopt.h>
#include <fcntl.h>
#include <netinet/in.h>
#include <libreswan.h>
#include <sys/utsname.h>

#include <arpa/nameser.h>
/* older versions lack ipseckey support */
#ifndef ns_t_ipseckey
# define ns_t_ipseckey 45
#endif

#include "constants.h"
#include "lswalloc.h"
#include "lswcrypto.h"
#include "lswlog.h"
#include "lswconf.h"
#include "secrets.h"
#include "mpzfuncs.h"

#include <nss.h>
#include <prerror.h>
#include <prinit.h>

char usage[] = "Usage: ipsec showhostkey [--ipseckey] | [--left ] | [--right ]\n"
             "                         [--precedence <precedence> ] [--gateway <gateway>] \n"
             "                         [--dump ] [--list ] \n"
             "                         [--dhclient ] [--file secretfile ] \n"
             "                         [--keynum count ] [--id identity ]\n"
             "                         [--rsaid keyid ] [--verbose] [--version]\n";

struct option opts[] = {
  {"help",	no_argument,	NULL,	'?',},
  {"left",	no_argument,	NULL,	'l',},
  {"right",	no_argument,	NULL,	'r',},
  {"dump",	no_argument,	NULL,	'D',},
  {"list",	no_argument,	NULL,	'L',},
  {"ipseckey",	no_argument,	NULL,	'K',},
  {"gateway",	required_argument,NULL,	'g',},
  {"precedence",required_argument,NULL,	'p',},
  {"dhclient",	no_argument,    NULL,	'd',},
  {"file",	required_argument,NULL,	'f',},
  {"keynum",	required_argument,NULL,	'n',},
  {"id",	required_argument,NULL,	'i',},
  {"rsaid",	required_argument,NULL,	'I',},
  {"version",	no_argument,	 NULL,	'V',},
  {"verbose",	no_argument,	 NULL,	'v',},
  {0,		0,	NULL,	0,}
};

char *progname = "ipsec showhostkey";	/* for messages */

void exit_tool(int code)
{
    tool_close_log();
    exit(code);
}

void
showhostkey_log(int mess_no, const char *message, ...)
{
    va_list args;

    va_start(args, message);
    vfprintf(stderr, message, args);
    va_end(args);
}


int print_key(struct secret *secret
	      , struct private_key_stuff *pks
	      , void *uservoid, bool disclose)
{
    int lineno = lsw_get_secretlineno(secret);
    struct id_list *l = lsw_get_idlist(secret);
    char idb[IDTOA_BUF];
    int count=1;

    char pskbuf[128];

    if(pks->kind == PPK_PSK || pks->kind==PPK_XAUTH) {
	datatot(pks->u.preshared_secret.ptr,
		pks->u.preshared_secret.len,
		'x', pskbuf, sizeof(pskbuf));
    }

    while(l) {
	idtoa(&l->id, idb, IDTOA_BUF);

	switch(pks->kind) {
	case PPK_PSK:
	    printf("%d(%d): PSK keyid: %s\n", lineno, count, idb);
	    if(disclose) printf("    psk: \"%s\"\n", pskbuf);
	    break;
	    
	case PPK_RSA:
	    printf("%d(%d): RSA keyid: %s with id: %s\n", lineno, count, pks->u.RSA_private_key.pub.keyid,idb);
	    break;
	    
	case PPK_XAUTH:
	    printf("%d(%d): XAUTH keyid: %s\n", lineno, count, idb);
	    if(disclose) printf("    xauth: \"%s\"\n", pskbuf);
	    break;
	    
	case PPK_PIN:
	    printf("%d:(%d) PIN key-type not yet supported for id: %s\n", lineno, count, idb);
	    break;
	}
	
	l=l->next;
	count++;
    }
    
    return 1;
}

int list_key(struct secret *secret,
	     struct private_key_stuff *pks,
	     void *uservoid)
{
    return print_key(secret, pks, uservoid, FALSE);
}

int dump_key(struct secret *secret,
	     struct private_key_stuff *pks,
	     void *uservoid)
{
    return print_key(secret, pks, uservoid, TRUE);
}


int pickbyid(struct secret *secret,
	     struct private_key_stuff *pks,
	     void *uservoid)
{
    char *rsakeyid = (char *)uservoid;

    if(strcmp(pks->u.RSA_private_key.pub.keyid, rsakeyid)==0) {
	return 0;
    }
    return 1;
}

struct secret *get_key_byid(struct secret *host_secrets, char *rsakeyid)
{
    return lsw_foreach_secret(host_secrets, pickbyid, rsakeyid);
}

void list_keys(struct secret *host_secrets)
{
    (void)lsw_foreach_secret(host_secrets, list_key, NULL);
}

char *get_default_keyid(struct secret *host_secrets)
{
    struct private_key_stuff *pks = lsw_get_pks(host_secrets);

    return pks->u.RSA_private_key.pub.keyid;
}

     
void dump_keys(struct secret *host_secrets)
{
    (void)lsw_foreach_secret(host_secrets, dump_key, NULL);
}

struct secret *pick_key(struct secret *host_secrets
			, char *idname)
{
    struct id id;
    struct secret *s;
    err_t e;

    e = atoid(idname, &id, FALSE);
    if(e) {
	printf("%s: key '%s' is invalid\n", progname, idname);
	exit(4);
    }

    s = lsw_find_secret_by_id(host_secrets, PPK_RSA
			      , &id, NULL, TRUE /* asymmetric */);
    
    if(s==NULL) {
	char abuf[IDTOA_BUF];
	idtoa(&id, abuf, IDTOA_BUF);
	printf("%s: can not find key: %s (%s)\n", progname, idname, abuf);
	exit(5);
    }

    return s;
}

unsigned char *pubkey_to_rfc3110(const struct RSA_public_key *pub,
				 unsigned int *keybuflen)
{
    unsigned char *buf;
    unsigned char *p;
    unsigned int elen;

    chunk_t e, n;


    e = mpz_to_n2(&pub->e);
    n = mpz_to_n2(&pub->n);
    elen = e.len;

    buf = alloc_bytes(e.len+n.len+3, "buffer for rfc3110");
    p = buf;
    
    if (elen <= 255)
	*p++ = elen;
    else if ((elen &~ 0xffff) == 0) {
	*p++ = 0;
	*p++ = (elen>>8) & 0xff;
	*p++ = elen & 0xff;
    } else {
	pfree(buf);
	return 0;	/* unrepresentable exponent length */
    }

    memcpy(p, e.ptr, e.len);
    p+=e.len;
    memcpy(p, n.ptr, n.len);
    p+=n.len;

    *keybuflen=(p-buf);

    return buf;
}    

void show_dnskey(struct secret *s
		 , char *idname
		 , int precedence
		 , char *gateway)
{
    char qname[256];
    char base64[8192];
    int gateway_type = 0;
    const struct private_key_stuff *pks = lsw_get_pks(s);
    unsigned char *keyblob;
    unsigned int keybloblen = 0;

    gethostname(qname, sizeof(qname));

    if(pks->kind != PPK_RSA) {
	printf("%s: wrong kind of key %s in show_dnskey. Expected PPK_RSA.\n",
		progname, enum_name(&ppk_names, pks->kind));
	exit(5);
    }

    keyblob = pubkey_to_rfc3110(&pks->u.RSA_private_key.pub, &keybloblen);

    datatot(keyblob, keybloblen, 's', base64, sizeof(base64));

	if(gateway != NULL){
		ip_address test;
		if(ttoaddr(gateway, strlen(gateway), AF_INET, &test) == NULL)
		   gateway_type = 1;
		else
		  if(ttoaddr(gateway, strlen(gateway), AF_INET6, &test) == NULL)
		     gateway_type = 2;
		  else {
			printf("%s: unknown address family for gateway %s", progname,gateway);
			exit(5);
		  }
	}
			
	printf("%s.    IN    IPSECKEY  %d %d 2 %s %s\n",
	       qname, precedence, gateway_type , (gateway == NULL) ? "." : gateway, base64);
}
     
void show_confkey(struct secret *s
		  , char *idname
		  , char *side)
{
    char base64[8192];
    const struct private_key_stuff *pks = lsw_get_pks(s);
    unsigned char *keyblob;
    unsigned int keybloblen = 0;

    if(pks->kind != PPK_RSA) {
        char *enumstr = "gcc is crazy";
        switch (pks->kind) {
        case PPK_PSK:
                enumstr = "PPK_PSK";
        case PPK_PIN:
                enumstr = "PPK_PIN";
        case PPK_XAUTH:
                enumstr = "PPK_XAUTH";
        default:
                sscanf(enumstr,"UNKNOWN (%d)",(int *)pks->kind);
        }
	printf("%s: wrong kind of key %s in show_confkey. Expected PPK_RSA.\n", progname,enumstr);
	exit(5);
    }

    keyblob = pubkey_to_rfc3110(&pks->u.RSA_private_key.pub, &keybloblen);

    datatot(keyblob, keybloblen, 's', base64, sizeof(base64));

    printf("\t# rsakey %s\n", 
	   pks->u.RSA_private_key.pub.keyid);
    printf("\t%srsasigkey=%s\n", side,
	   base64);
}
     


int main(int argc, char *argv[])
{
    char secrets_file[PATH_MAX];
    int opt;
    int errflg = 0;
    bool left_flg=FALSE;
    bool right_flg=FALSE;
    bool dump_flg=FALSE;
    bool list_flg=FALSE;
    bool ipseckey_flg=FALSE;
    bool dhclient_flg=FALSE;
    char *gateway = NULL;
    int precedence = 10;
    int verbose=0;
    const struct lsw_conf_options *oco = lsw_init_options();
    char *rsakeyid, *keyid;
    struct secret *host_secrets = NULL;
    struct secret *s;
    prompt_pass_t pass;

    rsakeyid=NULL;
    keyid=NULL;

    /* start up logging system */
    tool_init_log();

    snprintf(secrets_file, PATH_MAX, "%s/ipsec.secrets", oco->confdir);
    
    while ((opt = getopt_long(argc, argv, "", opts, NULL)) != EOF) {
	switch (opt) {
	case '?':
	    goto usage;
	    break;

	case 'l':
	    left_flg=TRUE;
	    break;
	case 'r':
	    right_flg=TRUE;
	    break;

	case 'D': /* --dump */
	    dump_flg=TRUE;
	    break;

	case 'K':
	    ipseckey_flg=TRUE;
	    gateway=clone_str(optarg, "gateway");
	    break;

	case 'p':
	    precedence = atoi(optarg);
	    if( (precedence < 0) || (precedence > 255)) {
		fprintf(stderr, "precedence must be between 0 and 255\n");
		exit(5);
	    }
	    break;
	case 'L':
	    list_flg=TRUE;
	    break;

	case 's':
	case 'R':
	case 'c':
	    break;

	case 'g':
	    ipseckey_flg=TRUE;
	    gateway=clone_str(optarg, "gateway");
	    break;

	case 'd':
	    break;

	case 'f':  /* --file arg */
	    secrets_file[0]='\0';
	    strncat(secrets_file, optarg, PATH_MAX-1);
	    break;

	case 'i':
	    keyid=clone_str(optarg, "keyname");
	    break;

	case 'I':
	    rsakeyid=clone_str(optarg, "rsakeyid");
	    break;

	case 'n':
	case 'h':
	    break;

	case 'v':
	    verbose++;
	    break;

	case 'V':
	    fprintf(stdout, "%s\n", ipsec_version_string());
	    exit(0);

	default:
	    goto usage;
	}
    }

    if(errflg) {
    usage:
	fputs(usage, stderr);
	exit(1);
    }
    
    if(!left_flg && !right_flg && !dump_flg && !list_flg
       && !ipseckey_flg && !dhclient_flg) {
	fprintf(stderr, "You must specify some operation\n");
	goto usage;
    }
    
    if((left_flg + right_flg + dump_flg + list_flg
	+ ipseckey_flg + dhclient_flg) > 1) {
	fprintf(stderr, "You must specify only one operation\n");
	goto usage;
    }

    if(verbose > 2) {
#ifdef DEBUG
	set_debugging(DBG_ALL);
#else
	fprintf(stderr, "verbosity cannot be set - compiled without DEBUG\n");
#endif
    }

    /* now load file from indicated location */
    pass.prompt=showhostkey_log;
    pass.fd = 2; /* stderr */

    PRBool nss_initialized = PR_FALSE;
    SECStatus rv;
    char buf[100];
    snprintf(buf, sizeof(buf), "%s",oco->confddir);
    loglog(RC_LOG_SERIOUS,"nss directory showhostkey: %s",buf);
    PR_Init(PR_USER_THREAD, PR_PRIORITY_NORMAL, 1);
    if ((rv = NSS_InitReadWrite(buf)) != SECSuccess) {
	fprintf(stderr, "%s: NSS_InitReadWrite returned %d\n",progname, PR_GetError());
	exit(1);
    }
   nss_initialized = PR_TRUE; 
   PK11_SetPasswordFunc(getNSSPassword); 

    load_lswcrypto();
    lsw_load_preshared_secrets(&host_secrets, verbose>0?TRUE:FALSE,
			       secrets_file, &pass);

    if (nss_initialized) {
	NSS_Shutdown();
    }
    PR_Cleanup();

    /* options that apply to entire files */
    if(dump_flg) {
	/* dumps private key info too */
	dump_keys(host_secrets);
	exit(0);
    }

    if(list_flg) {
	list_keys(host_secrets);
	exit(0);
    }

    if(rsakeyid) {
	printf("; picking by rsakeyid=%s\n", rsakeyid);
	s = get_key_byid(host_secrets, rsakeyid);
	keyid=rsakeyid;
    } else if(keyid) {
	printf("; picking by keyid=%s\n", keyid);
	s = pick_key(host_secrets, keyid);
    } else {
	/* Paul: This assumption is WRONG. Mostly I have PSK's above my
 	 * multiline default : RSA entry, and then this assumption breaks
 	 * The proper test would be for ": RSA" vs "@something :RSA"
 	 */
	/* default key is the *LAST* key, because it is first in the file.*/
	s=lsw_get_defaultsecret(host_secrets);
	keyid="default";
    }

    if(s==NULL) {
	printf("No keys found\n");
	exit(20);
    }

    if(left_flg) {
	show_confkey(s, keyid, "left");
	exit(0);
    }

    if(right_flg) {
	show_confkey(s, keyid, "right");
	exit(0);
    }

    if(ipseckey_flg) {
	show_dnskey(s, keyid,
		    precedence, gateway);
    }

    exit(0);
}

/*
 * Local Variables:
 * c-basic-offset:4
 * c-style: pluto
 * End:
 */
